src/osemgrep/cli/CLI.ml

```diff
@@ -137,22 +142,32 @@ let dispatch_subcommand argv =
         let subcmd_argv0 = argv0 ^ "-" ^ subcmd in
         subcmd_argv0 :: subcmd_args |> Array.of_list
       in
+      let experimental = Array.mem "--experimental" argv in
       (* coupling: with known_subcommands if you add an entry below.
        * coupling: with the main_help_msg if you add an entry below.
        *)
       match subcmd with
-      | "ci" -> Ci_subcommand.main subcmd_argv
-      | "install-semgrep-pro" -> missing_subcommand ()
-      | "login" -> Login_subcommand.main subcmd_argv
-      | "logout" -> Logout_subcommand.main subcmd_argv
-      | "lsp" -> Lsp_subcommand.main subcmd_argv
-      | "publish" -> missing_subcommand ()
-      | "scan" -> Scan_subcommand.main subcmd_argv
-      | "shouldafound" -> missing_subcommand ()
-      (* osemgrep-only *)
+      (* TODO: gradually remove those 'when experimental' guards as
+       * we progress in osemgrep port (or move the dispatch back
+       * to pysemgrep futher down, when we know we don't handle
+       * certain kind of arguments.
+       *)
+      | "ci" when experimental -> Ci_subcommand.main subcmd_argv
+      | "install-semgrep-pro" when experimental -> missing_subcommand ()
+      | "login" when experimental -> Login_subcommand.main subcmd_argv
+      | "logout" when experimental -> Logout_subcommand.main subcmd_argv
+      | "publish" when experimental -> missing_subcommand ()
+      | "scan" when experimental -> Scan_subcommand.main subcmd_argv
+      | "shouldafound" when experimental -> missing_subcommand ()
+      (* TODO: next target for not requiring the 'when experimental' guard! *)
+      | "lsp" when experimental -> Lsp_subcommand.main subcmd_argv
+      (* osemgrep-only: and by default! no need experimental! *)
       | "interactive" -> Interactive_subcommand.main subcmd_argv
       (* LATER: "dump", "test", "validate" *)
-      | _else_ -> (* should have defaulted to 'scan' above *) assert false)
+      | _else_ ->
+          if experimental (* should have defaulted to 'scan' above *) then
+            assert false
```

Comment:
Can we make this a better error message than assert false since this is now exposed to end users?